arngll/src/frame_info.rs

```diff
@@ -0,0 +1,664 @@
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use std::fmt::{Debug, Formatter};
+use super::*;
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct NetworkId(pub u16);
+
+impl NetworkId {
+    pub fn from_iter<'a, T: Iterator<Item=&'a u8>>(iter: &mut T) -> NetworkId {
+        let msb = *iter.next().unwrap();
+        let lsb = *iter.next().unwrap();
+        NetworkId(((msb as u16)<<8) | (lsb as u16))
+    }
+}
+
+impl Debug for NetworkId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[{:04X}]", self.0)
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum FrameType {
+    Beacon,
+    Data,
+    Ack,
+    MacCommand,
+}
+
+impl FrameType {
+    pub fn try_from_u8(x: u8) -> Option<FrameType> {
+        match x {
+            0 => Some(Self::Beacon),
+            1 => Some(Self::Data),
+            2 => Some(Self::Ack),
+            3 => Some(Self::MacCommand),
+            _ => None,
+        }
+    }
+
+    pub fn to_u8(&self) -> u8 {
+        match self {
+            Self::Beacon => 0,
+            Self::Data => 1,
+            Self::Ack => 2,
+            Self::MacCommand => 3,
+        }
+    }
+}
+
+impl TryFrom<u8> for FrameType {
+    type Error = anyhow::Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        FrameType::try_from_u8(value).ok_or(format_err!("{} is not a valid frame type", value))
+    }
+}
+
+/// Enum encoding the length of a Message Integrity Code.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum MicLen {
+    Mic32 = 0,
+    Mic64 = 1,
+    Mic96 = 2,
+    Mic128 = 3,
+}
+
+impl MicLen {
+    pub fn try_from_u8(x: u8) -> Option<MicLen> {
+        match x {
+            0 => Some(Self::Mic32),
+            1 => Some(Self::Mic64),
+            2 => Some(Self::Mic96),
+            3 => Some(Self::Mic128),
+            _ => None,
+        }
+    }
+
+    pub fn to_u8(&self) -> u8 {
+        match self {
+            Self::Mic32 => 0,
+            Self::Mic64 => 1,
+            Self::Mic96 => 2,
+            Self::Mic128 => 3,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Mic32 => 4,
+            Self::Mic64 => 8,
+            Self::Mic96 => 12,
+            Self::Mic128 => 16,
+        }
+    }
+}
+
+impl TryFrom<u8> for MicLen {
+    type Error = anyhow::Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        MicLen::try_from_u8(value).ok_or(format_err!("{} is not a valid MIC length", value))
+    }
+}
+
+/// Message Integrity Code.
+#[derive(Clone, Eq, PartialEq)]
+pub struct Mic {
+    pub len: MicLen,
+    pub code: [u8; 16],
```

Comment:
Given that this is `pub`, what prevents software from reaching in and grabbing zero bytes after the code?

I wonder if, perhaps, a better organization would be to make Mic an enum, with variants that carry an explicitly sized array?

    pub enum Mic {
        Mic32([u8; 4]),
        Mic64([u8; 8]),
        Mic96([u8; 12]),
        Mic128([u8; 16]),
    }

Etc....

Reply:
Good idea, will do in a later commit.